src/VisualStudio/Core/Def/Implementation/LanguageClient/VisualStudioLspSolutionProvider.cs

```diff
@@ -3,45 +3,53 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
-using System.Collections.Immutable;
 using System.Composition;
+using System.Linq;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host.LanguageServerProtocol;
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.LanguageServer;
-using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageClient
 {
     [Export(typeof(ILspSolutionProvider)), Shared]
     internal class VisualStudioLspSolutionProvider : ILspSolutionProvider
     {
-        private readonly VisualStudioWorkspace _visualStudioWorkspace;
-        private readonly MiscellaneousFilesWorkspace _miscellaneousFilesWorkspace;
+        private readonly ILspWorkspaceRegistrationService _registrationService;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public VisualStudioLspSolutionProvider(VisualStudioWorkspace visualStudioWorkspace, MiscellaneousFilesWorkspace miscellaneousFilesWorkspace)
+        public VisualStudioLspSolutionProvider(ILspWorkspaceRegistrationService registrationService)
         {
-            _visualStudioWorkspace = visualStudioWorkspace;
-            _miscellaneousFilesWorkspace = miscellaneousFilesWorkspace;
+            _registrationService = registrationService;
         }
 
-        public Solution GetCurrentSolutionForMainWorkspace()
+        public (DocumentId?, Solution) FindDocumentAndSolution(TextDocumentIdentifier? textDocument, string? clientName)
         {
-            return _visualStudioWorkspace.CurrentSolution;
-        }
+            // Assume the first workspace registered is the main one
```

Comment:
>Assume the first workspace registered is the main one [](start = 15, length = 53)

Seems like a fragile assumption.

Reply:
Yes, I agree, but I also don't know that having an explicit `SetMainWorkspace` makes sense either. If the AnyCodeWorkspace somehow gets activated when the RoslynVisualStudioWorkspace is already active, I wouldn't want it to become the main workspace, but in Any Code scenarios you would want it to be the main one. Maybe that scenario never happens though?

Reply:
At this point, which requests _don't_ have a document path? If it's things like symbol search, I'd expect the answer is they get the list of CurrentSolutions to look at and there's no longer a concept of a privileged "main" solution; instead they're searching all solutions and merging the results. If we have a case where we truly cannot process multiple solutions _and_ we don't have a document path, then we have a fundamental design issue that we need to readdress.

Reply:
> At this point, which requests don't have a document path?

That's a very good question. The main two are workspace symbols and workspace pull diagnostics. I'm not sure we want those pulling from all workspaces though, as that would see Metadata as Source for example, which is probably not something we want diagnostics for, right?